✨ Added specific newsletter support for bulk unsubscribes

ronaldlangeveld opened the pull request:
ref https://github.com/TryGhost/Team/issues/2013

Added support to be able to bulk unsubscribe a selected (filtered) list on members from specific newsletters.

- [x] There's a clear use-case for this code change, explained below
- [x] Commit message has a short title & references relevant issues
- [x] The build will pass (run `yarn test:all` and `yarn lint`)

ronaldlangeveld committed:
in progress latest

ronaldlangeveld committed:
Added newsletter dropdown selection
ref https://github.com/TryGhost/Team/issues/2013

- Added function to select a specific newsletter if > 1 newsletter for bulk unsubscriptions

ronaldlangeveld committed:
Added api support for bulk 'specific' newsletter unsubscribe
ref https://github.com/TryGhost/Team/issues/2013

- takes a newsletter id argument into account when processing the bulk unsubscribe request to know which newsletter should be unsubscribed
- added a bulkFind bulk-operation - there's probably existing helpers around the codebase, but small implimentation.
- added e2e test to test if you can unsubscribe multiple members from a specific newsletter.

ronaldlangeveld committed:
Fixed linting
ref https://github.com/TryGhost/Team/issues/2013

- Fixed linting errors

allouis reviewed (commented):
I think the general approach makes sense, but I think we should consider creating a `MemberNewsletter` bookshelf model 🤔 

That way we can use `findAll` rather than `bulkFind` and then we can use `bulkDelete` method to delete all the records.

ronaldlangeveld committed:
Removed auto-imported import by vscode
ref https://github.com/TryGhost/Team/issues/2013

ronaldlangeveld committed:
Added new `MemberNewsletter` bookshelf model
ref https://github.com/TryGhost/Ghost/pull/15742#pullrequestreview-1171367626 https://github.com/TryGhost/Team/issues/2013

- Added a new bookshelf model that queries the `MemberNewsletters` table.
- Changed the unsubscribe actions to a single reduce to avoid double looping. ref https://github.com/TryGhost/Ghost/pull/15742/files/4538e2f86fc1891d44ebb5268dcecb6ea0f6652d#r1016097188

ronaldlangeveld committed:
Merge branch 'main' into bulk-newsletter-unsub

ronaldlangeveld committed:
Cleanup unused code & fixed linting

kevinansfield commented on ghost/admin/app/components/members/modals/bulk-unsubscribe.hbs:
```diff
@@ -43,11 +43,28 @@
                         <GhLoadingSpinner />
                     {{else}}
+                        {{#if this.hasMultipleNewsletters }}
+                         <p>
+                            Which newsletter should these
+                            <strong data-test-text="unsubscribe-count">{{gh-pluralize countFetcher.count "member"}}</strong> be unsubscribed from?
+                        </p>
+                        <OneWaySelect
+                            @value={{this.selectedNewsletterId}}
+                            @options={{this.newsletterList}}
+                            @optionLabelPath="name"
+                            @optionValuePath="value"
+                            @optionTargetPath="value"
+                            @update={{this.setSelectedNewsletter}}
+                            data-test-select="members-single-newsletter-unsubscribe"
+                        />
```
Small thing but it's good to always indent html to match the conditionals, helps to keep the code understandable at a glance and avoid problems like we saw in the flag cleanup where the lack of proper indentation made it tricky to see which blocks should be cleaned up.

kevinansfield commented on ghost/admin/app/components/members/modals/bulk-unsubscribe.js:
```diff
@@ -19,27 +22,62 @@ export default class BulkUnsubscribeMembersModal extends Component {
+            name: 'All newsletters',
+            value: 'all'
+        }];
+        activeNewsletters.map((newsletter) => {
```
If you're not using the return value of a map please use `forEach` instead.

kevinansfield reviewed (approved):
A couple of style related comments to get in the habit of doing but otherwise the Ember stuff LGTM 👍

ronaldlangeveld committed:
Merge branch 'main' into bulk-newsletter-unsub

ronaldlangeveld committed:
Switch map to forEach
ref https://github.com/TryGhost/Team/issues/2013

allouis commented on ghost/members-api/lib/repositories/member.js:
```diff
@@ -726,9 +729,21 @@ module.exports = class MemberRepository {
-            return await this._Member.bulkDestroy(memberIds, 'members_newsletters', {column: 'member_id'});
+            const hasNewsletterSelected = (Object.prototype.hasOwnProperty.call(data, 'newsletter') && data.newsletter !== null);
+            if (hasNewsletterSelected) {
+                const newsletters = await this._MemberNewsletter.where('newsletter_id', data.newsletter).fetchAll();
```
Rather than use `reduce` below to get the correct list we should be able to make an SQL query to get the correct data.

So could do something like `this._MemberNewsletter.where().andWhereIn()` Or we can use the `getFilteredCollectionQuery` method above and pass in an NQL filter like `newsletter_id:blah+member_id:[id1,id2,id3...]`

allouis commented on ghost/members-api/lib/repositories/member.js:
```diff
@@ -726,9 +729,21 @@ module.exports = class MemberRepository {
+                    }
+                    return acc;
+                }, []);
+                return await this._Member.bulkDestroy(toUnsubscribe, 'members_newsletters', {column: 'id'});
```
This can be written as `this._MemberNewsletter.bulkDestroy(toUnsubscribe)`

allouis commented on ghost/members-api/lib/repositories/member.js:
```diff
@@ -726,9 +729,21 @@ module.exports = class MemberRepository {
+                return await this._Member.bulkDestroy(toUnsubscribe, 'members_newsletters', {column: 'id'});
+            }
+            if (!hasNewsletterSelected) {
+                return await this._Member.bulkDestroy(memberIds, 'members_newsletters', {column: 'member_id'});
```
What would be really cool is if we could update the `bulkDestroy` method to make calls like this nicer, imagine if we could instead do:
```js
this._MemberNewsletter.bulkDestroy({member_id: memberIds});


// bulk-operations.js

async bulkDestroy(data, tableName, options) {
  if (!Array.isArray(data)) {
    const [key, ...keys] = Object.keys(data);

    assert(keys.length === 0);
    return this.bulkDestroy(data[key], tableName, {column: key});
  }
  // ...
}
```

Not necessarily in scope for this PR but would be much nicer!

allouis reviewed (commented):
Looks good just a couple comments!

ronaldlangeveld committed:
Removed defaults method

ronaldlangeveld committed:
Replaced fetchAll query with nql filter.
ref https://github.com/TryGhost/Ghost/pull/15742#discussion_r1022240819

ronaldlangeveld committed:
Merge branch 'main' into bulk-newsletter-unsub
